src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -967,11 +972,14 @@ protected override BoundNode VisitReturnStatementNoAdjust(BoundReturnStatement n
         private TypeSymbolWithAnnotations GetReturnType()
         {
             var method = (MethodSymbol)_member;
-            var returnType = (_useMethodSignatureReturnType ? _methodSignatureOpt : method).ReturnType;
+            var returnType = (_useMethodSignatureReturnType ? _methodSignatureOpt : method).ReturnType.SetPossiblyNullableReferenceTypeTypeParametersAsNullable();
             Debug.Assert((object)returnType != LambdaSymbol.ReturnTypeIsBeingInferred);
-            return method.IsGenericTaskReturningAsync(compilation) ?
-                ((NamedTypeSymbol)returnType.TypeSymbol).TypeArgumentsNoUseSiteDiagnostics.Single() :
-                returnType;
+            if (method.IsGenericTaskReturningAsync(compilation))
+            {
+                returnType = ((NamedTypeSymbol)returnType.TypeSymbol).TypeArgumentsNoUseSiteDiagnostics.Single().SetPossiblyNullableReferenceTypeTypeParametersAsNullable();
```

Comment:
>SetPossiblyNullableReferenceTypeTypeParametersAsNullable [](start = 113, length = 56)

I'm not sure if we need this second call. `returnType` should already have been "cleansed", including its type arguments.

Reply:
> I'm not sure if we need this second call. returnType should already have been "cleansed", including its type arguments.

That is correct. I am planning to get rid of all the SetPossiblyNullableReferenceTypeTypeParametersAsNullable cleansing everywhere in my next commit.

---
In reply to: [230531004](https://github.com/dotnet/roslyn/pull/30913#discussion_r230531004) [](ancestors = 230531004)